### ISSUE TITLE
loading model does not initialize svc_type, kernel_type

### ISSUE THREAD

**Issue by denimboy** [rosasaul/Algorithm-SVM#2]
Create a simple model for exclusive or like so:

``` shell
    $ cat xor.dat
    0 0:0 1:0
    1 0:0 1:1
    1 0:1 1:0
    0 0:1 1:1
    $ svm-train -t 0 -s 1 xor.dat xor.model
    ...
    $ head -2 xor.model
    svm_type nu_svc
    kernel_type linear
```

Notice we created a Nu-SVC style clasifier with a linear kernel.  Now lets load that model back into libsvm via perl and check:

``` perl
my $svm = Algorithm::SVM->new();
$svm->load('xor.model') || die "error loading model";
print "svm_type ", $svm->svm_type(), "\n";
print "kernel_type ", $svm->kernel_type(), "\n";
```

The above outputs:

``` shell
   svm_type C-SVC
   kernel_type radial
```

which are default values and not what was in the model file.  This makes me worry that the model is not properly loaded.

Setting the attribute and then accessing it returns the value just set:

``` perl
$svm->kernel_type('linear');
print "kernel_type ", $svm->kernel_type(), "\n";
```

but I am unsure at this point what SVM type or kernel type I am actually working with; the one specified in the model or the one I just set via the setter method.

**Comment by rosasaul** [comment]
Yeah that's bad design. There's a mess of issues in the way the C code was arranged which I'd meant to fix.

This issue you're seeing is that the file loading is occurring in the C code, but perl has a copy of the instantiated param object with values in it in addition to the one defined in C. When you poke at the values directly you get what perl still thinks is the values but the values in the C memory have been updated by the file upload svm_load_model(). So you'd need to request the type via a method call and build a method that asks the C for the current value of said variables, that or update the perl load method to check on the variables post loading the model.

I moved away from using this because of some of these issues, and use one that requires a bit more direct C manipulation (you need to request memory as well as destroy it from inside perl) but doesn't leave you guessing on issues like this. https://github.com/rosasaul/libsvm-perl

**Comment by denimboy** [comment]
Hi,

On Dec 2, 2014, at 12:16 PM, Saul Rosa notifications@github.com wrote:

> Yeah that's bad design. There's a mess of issues in the way the C code was arranged which I'd meant to fix.
> 
> This issue you're seeing is that the file loading is occurring in the C code, but perl has a copy of the instantiated param object with values in it in addition to the one defined in C. When you poke at the values directly you get what perl still thinks is the values but the values in the C memory have been updated by the file upload svm_load_model(). So you'd need to request the type via a method call and build a method that asks the C for the current value of said variables, that or update the perl load method to check on the variables post loading the model.

So can I assume that the C has loaded the model properly and will predict classes as expected?

Problem is I have been building models using svm-train then loading these in Algorithm::SVM and predicting but the results look wrong.  I started back peddling to try and pin point an error which is how I got here.

> I moved away from using this because of some of these issues, and use one that requires a bit more direct C manipulation (you need to request memory as well as destroy it from inside perl) but doesn't leave you guessing on issues like this. https://github.com/rosasaul/libsvm-perl

Ah, swig wrapper?  That should be very similar to the types python wrapper that ships with libsvm.

I see you have another LIBSVM based project libml-perl.  Did the two have different goals?  Looks a little like you started libml-perl but gave up on it for the simple swig based libsvm-perl.
